### ISSUE TITLE
Upgrading specific dependencies -- will not upgrade to latest versions

### ISSUE THREAD

**Issue by jonasfj** [dart-lang/site-www#3724]
I think the semantics covered here: https://dart.dev/tools/pub/cmd/pub-upgrade#upgrading-specific-dependencies

Is incorrect, it used to be that we would upgrade to the latest version available. But afaik now `dart pub upgrade foo bar` just means that only `foo` and `bar` is upgraded. Both are upgraded as far as possible without violating constraints in `pubspec.yaml` and without upgrading other dependencies.

That's `foo` and `bar` are unlock from the `pubspec.lock`. Maybe it not necessary to explain that, but the latest version semantics have changed.

![image](https://user-images.githubusercontent.com/149732/145993261-7cce162b-cf16-4aed-96a2-6e4eb723a6a3.png)

**Comment by atsansone** [comment]
@jonasfj : Can you clarify the resolution you seek?

Do we need to change the text from:
```markdown
Upgrading a dependency upgrades its transitive dependencies to their latest versions as well. Usually, no other dependencies are updated; they stay at the versions that are locked in the lockfile. However, if the requested upgrades cause incompatibilities with these locked versions, they are selectively unlocked until a compatible set of versions is found.
```
to something like:
```markdown
Listing specific dependencies upgrades only those listed. In the previous example, `dart pub upgrade` upgrades both `test` and `args` to the latest version that doesn't violate constraints in the `pubspec.yaml` file. It doesn't upgrade other dependencies. If the requested upgrades cause incompatibilities with these locked versions of dependencies, `dart pub` unlocks those dependencies until it finds a compatible set of versions.
```

**Comment by jonasfj** [comment]
Yeah, I think the word "latest version" is wrong. I mean it communicate an intuition, but it's not the semantics.

    ## Upgrading specific dependencies
    
    You can tell `dart pub upgrade` to upgrade specific dependencies while leaving the rest
    of the dependencies alone (if possible). For example:
    
    ```console
    $ dart pub upgrade test args
    Dependencies upgraded!
    ```

    This will attempt to upgrade the `test` and `args` dependencies, while leaving
    other dependencies locked to the version in the lockfile. The other dependencies
    will only be unlocked, if this is necessary to satisfy constraints in `pubspec.yaml`.
    
    This means that upgrading a specific dependency does not by default upgrade its
    transitive dependencies. Nor does it necessarily upgrade the dependency, if a
    newer version requires transitive dependencies to be upgraded.
    
    To upgrade a specific dependency and all its transitive dependencies use
    the `--unlock-transitive` flag.
    
    ```console
    $ dart pub upgrade --unlock-transitive test args
    ```

The word "latest version" here is almost always wrong. Because a package very often has a constraint specified in `pubspec.yaml`.

For example:
```yaml
name: myapp
dependencies:
  args: ^1.2.0 # this means >= 1.2.0 < 2.0.0
```

So if you do `dart pub upgrade` (to upgrade all dependencies, including transitive dependencies), or `dart pub upgrade args` or `dart pub upgrade --unlock-transitive args`, you will still NEVER get the latest version of `args`, if package `args` has a version `2.0.0` or higher.

I realize that the language I'm proposing is a bit vague, but that's because the behavior is vague.

We can't really say:
> "You can tell dart pub upgrade to upgrade specific dependencies to the **latest version**" 

because it's just plain wrong 🤣 

I think we should also be careful to avoid implications of optimality with respect to any criteria:
> while leaving the rest of the dependencies alone **as much as possible**

Because while the constraint solver is good and smart, it does not explore the entire state-space. I think it would be cool if the constraint solver did a best-first exploration of the state-space so that we could promise that the solution is optimal with respect to some definition of _best solution_.

But the constraint solver does a depth-first exploration with local preferences for picking higher versions. This means that the solver can go down a sub-optimal branch where it ends up with a solution that has many older versions, even though a possible to find a solution that have newer versions.

I don't think we should explain these subtleties in the documentation. We should just avoid making or implying things that aren't true. Which means it's better to vague about the behavior.

In practice:
 * It's rare that the solver doesn't do the right thing. It's heuristics are decent, and most importantly picking a new version of package A usually also makes it easier to pick a newer version of package B. So the kind of constraint that could cause it to go down a sub-optimal path are rare.
 * If a users really wants to get version `1.2.3` of package `foo`, then said user can just write: `foo: ^1.2.3` in the `pubspec.yaml` which will force the solver to try this. And if it can't do it, it'll explain why it can't.

So while I think a best-first exploration would be cool. It would probably eat up non-trivial memory in the rare degenerate case, and it would be probably be exceedingly rare that it matters whether you did a best-first exploration or a depth-first exploration.

Finally, there is the problem that the best-first exploration requires us to define what is "best". And it's totally possible to have a scenario where you can't get the latest version of package A and B, but only one of them -- thus, defining what "best" means is even hard, and explaining it to users would be difficult too.

I'll stop rambling, all I'm really suggesting is that we are a bit more vague here, so users don't feel like we promised to get them the latest version, or do something as much as possible 🤣
